### PR TITLE
Prevent loading of incompatible native modules

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alanjo@microsoft.com>

Resolves: #1964 

## Description

Add a size field to metadata_table_t to detect mismatch between bpf2c and eBPF for Windows runtime.

## Testing

CI/CD

## Documentation

No.

### CODE REVIEW

File: include/bpf2c.h [hunk 1]
```diff
@@ -53,62 +59,106 @@ extern "C"
...
+        ebpf_program_type_t* program_type;        ///< Type of the program.
+        ebpf_attach_type_t* expected_attach_type; ///< Expected attach type of the program.
+        const uint8_t* program_info_hash;         ///< Hash of the program info.
+        size_t program_info_hash_length;          ///< Length of the program info hash.
     } program_entry_t;
 
+    /**
+     * @brief Version information for the bpf2c compiler.
+     * This structure contains the version information for the bpf2c compiler that generated the module. It can be
+     * used to determine if the module is compatible with the current version of eBPF for Windows runtime.
```
Comment (dthaler) [inline include/bpf2c.h:103]:
```suggestion
     * used to determine if the module is compatible with the current version of the eBPF for Windows runtime.
```

File: include/bpf2c.h [hunk 2]
```diff
@@ -53,62 +59,106 @@ extern "C"
...
+     */
     typedef struct _metadata_table
     {
-        void (*programs)(_Outptr_result_buffer_maybenull_(*count) program_entry_t** programs, _Out_ size_t* count);
-        void (*maps)(_Outptr_result_buffer_maybenull_(*count) map_entry_t** maps, _Out_ size_t* count);
-        void (*hash)(_Outptr_result_buffer_maybenull_(*size) const uint8_t** hash, _Out_ size_t* size);
+        size_t size; ///< Size of this structure. Used for versioning.
+        void (*programs)(
+            _Outptr_result_buffer_maybenull_(*count) program_entry_t** programs,
+            _Out_ size_t* count); ///< Returns the list of programs in the this module.
```
Comment (dthaler) [inline include/bpf2c.h:122]:
```suggestion
            _Out_ size_t* count); ///< Returns the list of programs in this module.
```

File: include/bpf2c.h [hunk 3]
```diff
@@ -53,62 +59,106 @@ extern "C"
...
+        void (*maps)(
+            _Outptr_result_buffer_maybenull_(*count) map_entry_t** maps,
+            _Out_ size_t* count); ///< Returns the list of maps in the this module.
```
Comment (dthaler) [inline include/bpf2c.h:125]:
```suggestion
            _Out_ size_t* count); ///< Returns the list of maps in this module.
```

File: include/bpf2c.h [hunk 4]
```diff
@@ -53,62 +59,106 @@ extern "C"
...
+        void (*hash)(
+            _Outptr_result_buffer_maybenull_(*size) const uint8_t** hash,
+            _Out_ size_t* size); ///< Returns the hash of ELF file used to generate this module.
```
Comment (dthaler) [inline include/bpf2c.h:128]:
```suggestion
            _Out_ size_t* size); ///< Returns the hash of the ELF file used to generate this module.
```

**Review by dthaler — APPROVED** [review]

**Review by dthaler — DISMISSED** [review]

File: libs/execution_context/ebpf_native.c [hunk 5]
```diff
@@ -364,6 +364,18 @@ _ebpf_native_provider_attach_client_callback(
...
+            EBPF_TRACELOG_LEVEL_ERROR,
+            EBPF_TRACELOG_KEYWORD_NATIVE,
+            "The metadata table size is wrong for client module. The version of bpf2c used to generate this module "
+            "may be too old.",
+            *client_module_id);
+        goto Done;
+    }
+
     bpf2c_version_t client_version = {0, 0, 0};
     table->version(&client_version);
```
Comment (saxena-anurag) [inline libs/execution_context/ebpf_native.c:380]: nit: Even though we have added check for metadata table size, should we add a NULL check here for `version` (and for other function pointers also in the metadata table)

**Review by saxena-anurag — DISMISSED** [review]

File: libs/execution_context/ebpf_native.c (hunk 5 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_native.c:380]: Why? In what scenario would bpf2c emit a struct with null in it?

Comment (saxena-anurag) [reply libs/execution_context/ebpf_native.c:380]: possibly due to a bug? This is just a suggestion though.

**Review by dthaler — APPROVED** [review]

**Pull request merged**